src/Interpreters/Cluster.cpp

```diff
@@ -838,14 +840,21 @@ Cluster::Cluster(Cluster::ReplicasAsShardsTag, const Cluster & from, const Setti
         std::shuffle(shards_info.begin(), shards_info.end(), gen);
         shards_info.resize(max_hosts);
 
+        Addresses all_addresses_;
```

Comment:
Can you put the entire logic that lives inside this if statement in a separate method? IMHO, it is now slightly more complex and it deserves a named method. Perhaps something like:

```
static void constrainShardInfoAndAddressesToMaxHosts(ShardInfo & shard_info, Addresses & addresses)
{
        pcg64_fast gen{randomSeed()};
        std::shuffle(shards_info.begin(), shards_info.end(), gen);
        shards_info.resize(max_hosts);
        
        Addresses shuffled_addresses;
        for (auto & shard_info : shards_info)
        {
            shuffled_addresses.emplace_back(addresses[shard_info.shard_num - 1]);
            shard_info.shard_num = ++shard_num;
        }

         addresses.swap(shuffled_addresses);
}
```